packages/client/src/index.ts

```diff
@@ -0,0 +1,378 @@
+import {
+  CodeTokenResponse,
+  decodeIdToken,
+  fetchOidcConfig,
+  fetchTokenByAuthorizationCode,
+  fetchTokenByRefreshToken,
+  generateCodeChallenge,
+  generateCodeVerifier,
+  generateSignInUri,
+  generateSignOutUri,
+  generateState,
+  IdTokenClaims,
+  Prompt,
+  Requester,
+  revoke,
+  verifyAndParseCodeFromCallbackUri,
+  verifyIdToken,
+  withReservedScopes,
+} from '@logto/js';
+import { Nullable } from '@silverhand/essentials';
+import { createRemoteJWKSet } from 'jose';
+import once from 'lodash.once';
+import { assert, Infer, string, type } from 'superstruct';
+
+import { LogtoClientError } from './errors';
+import { Storage } from './storage';
+import { buildAccessTokenKey, getDiscoveryEndpoint } from './utils';
+
+export type { IdTokenClaims, LogtoErrorCode } from '@logto/js';
+export { LogtoError, OidcError, Prompt } from '@logto/js';
+export * from './errors';
+
+export type LogtoConfig = {
+  endpoint: string;
+  appId: string;
+  scopes?: string[];
+  resources?: string[];
+  prompt?: Prompt;
+  usingPersistStorage?: boolean;
+};
+
+export type AccessToken = {
+  token: string;
+  scope: string;
+  expiresAt: number; // Unix Timestamp in seconds
+};
+
+export const LogtoSignInSessionItemSchema = type({
+  redirectUri: string(),
+  codeVerifier: string(),
+  state: string(),
+});
+
+export type LogtoSignInSessionItem = Infer<typeof LogtoSignInSessionItemSchema>;
+
+export type HandleRedirect = (url: string) => void;
+
+export default class LogtoClient {
+  protected readonly logtoConfig: LogtoConfig;
+  protected readonly getOidcConfig = once(this._getOidcConfig);
+  protected readonly getJwtVerifyGetKey = once(this._getJwtVerifyGetKey);
+
+  protected readonly storage: Storage;
+  protected readonly requester: Requester;
+  protected readonly handleRedirect: HandleRedirect;
+
+  protected readonly accessTokenMap = new Map<string, AccessToken>();
+
+  private readonly getAccessTokenPromiseMap = new Map<string, Promise<string>>();
+  private _idToken: Nullable<string>;
+
+  constructor(
+    logtoConfig: LogtoConfig,
+    {
+      requester,
+      storage,
+      handleRedirect,
+    }: { requester: Requester; storage: Storage; handleRedirect: HandleRedirect }
```

Comment:
nit: Maybe we should include logtoConfig in this new object as well? And it would be better to define a type name for it

Reply:
I am gonna name it as something like "adapters", so I put logtoConfig outside the object. Good call, I'll think about it after finish crypto things.